File: src/CSES.Solutions/BaseSolverTests.cs
using CSES.Core;
using Xunit;

namespace CSES.Solutions;

/// <summary>
/// Base class for all solver tests. Provides helper methods to run tests
/// against input/output file pairs.
/// </summary>
public abstract class BaseSolverTests<TSolver> where TSolver : ISolver, new()
{
    protected readonly TSolver Solver;

    protected BaseSolverTests()
    {
        Solver = new TSolver();
    }

    /// <summary>
    /// Runs a test by reading the input from a file, solving it,
    /// and comparing the result with the expected output file.
    /// </summary>
    /// <param name="testDataFolder">The folder containing the test data (relative to the test class)</param>
    /// <param name="testNumber">The test number (e.g., 1 for 1.in/1.out)</param>
    protected void RunTest(string testDataFolder, int testNumber)
    {
        var inputPath = Path.Combine(testDataFolder, $"{testNumber}.in");
        var outputPath = Path.Combine(testDataFolder, $"{testNumber}.out");

        Assert.True(File.Exists(inputPath), $"Input file not found: {inputPath}");
        Assert.True(File.Exists(outputPath), $"Output file not found: {outputPath}");

        var input = File.ReadAllText(inputPath);
        var expectedOutput = File.ReadAllText(outputPath).TrimEnd('\r', '\n');
        var actualOutput = Solver.Solve(input).TrimEnd('\r', '\n');

        Assert.Equal(expectedOutput, actualOutput);
    }

    /// <summary>
    /// Gets all test numbers available in the test data folder.
    /// </summary>
    protected static IEnumerable<object[]> GetTestNumbers(string testDataFolder)
    {
        if (!Directory.Exists(testDataFolder))
        {
            yield break;
        }

        var inputFiles = Directory.GetFiles(testDataFolder, "*.in")
            .Select(Path.GetFileNameWithoutExtension)
            .Where(name => int.TryParse(name, out _))
            .Select(int.Parse)
            .OrderBy(n => n);

        foreach (var testNumber in inputFiles)
        {
            yield return new object[] { testNumber };
        }
    }
}

File: src/CSES.FSharp.Solutions/BaseSolverTests.fs
namespace CSES.FSharp.Solutions

open System
open System.IO
open CSES.Core
open Xunit

/// Base class for all F# solver tests. Provides helper methods to run tests
/// against input/output file pairs.
[<AbstractClass>]
type BaseSolverTests<'TSolver when 'TSolver :> ISolver and 'TSolver : (new : unit -> 'TSolver)>() =
    let solver = new 'TSolver()

    member _.Solver = solver

    /// Runs a test by reading the input from a file, solving it,
    /// and comparing the result with the expected output file.
    member _.RunTest(testDataFolder: string, testNumber: int) =
        let inputPath = Path.Combine(testDataFolder, $"{testNumber}.in")
        let outputPath = Path.Combine(testDataFolder, $"{testNumber}.out")

        Assert.True(File.Exists(inputPath), $"Input file not found: {inputPath}")
        Assert.True(File.Exists(outputPath), $"Output file not found: {outputPath}")

        let input = File.ReadAllText(inputPath)
        let expectedOutput = File.ReadAllText(outputPath).TrimEnd([| '\r'; '\n' |])
        let actualOutput = solver.Solve(input).TrimEnd([| '\r'; '\n' |])

        Assert.Equal(expectedOutput, actualOutput)

/// Module with helper functions for test data discovery
module TestHelpers =
    /// Gets all test numbers available in the test data folder as xUnit theory data.
    let getTestNumbers (testDataFolder: string) : seq<obj[]> =
        if not (Directory.Exists(testDataFolder)) then
            Seq.empty
        else
            Directory.GetFiles(testDataFolder, "*.in")
            |> Seq.map Path.GetFileNameWithoutExtension
            |> Seq.choose (fun name ->
                match Int32.TryParse(name) with
                | true, n -> Some n
                | false, _ -> None)
            |> Seq.sort
            |> Seq.map (fun n -> [| box n |])
